ztp/extra-manifests-builder/01-disk-encryption-pcr-rebind/luks-helpers.sh

```diff
@@ -0,0 +1,159 @@
+    return $FALSE
+}
+
+# create a temporary key in the reserved slot to disable PCR protection
+addReservedSlot() {
+    logInfo "reservedSlotPresent=$1 device=$2 slot=$3 with PCR IDs=$4 and clevis config=$5"
+    if [ $1 == $TRUE ]; then
+        logInfo "reserve slot already present, no need to add again"
+        clevis luks list -d $2 || true
+        return
+    fi
+    logInfo "adding reserved slot on device=$device"
+    ANYPASS=$(openssl rand -base64 21)
+    echo -e "$ANYPASS\n" | clevis luks bind -s $RESERVED_SLOT -d $2 tpm2 '{}' || true
+    clevis luks list -d $2 || true
+}
+
+# remove the temporary key in the reserved slot to enable PCR protection
+removeReservedSlot() {
+    logInfo "removing luks reserved slot 31 in disk $1"
+    echo "clevis luks unbind -s $RESERVED_SLOT -d $1 -f" | sed 's@/@\/@g' | bash || true
+}
+
+#gets the list of luks devices in the system
+getLUKSDevices() {
+    results=$(lsblk --fs -l | grep crypto_LUKS | awk '{printf "/dev/" $1 "|"}')
+    logDebug "got luks devices across all drives: $results"
+    echo $results
+}
+
+# create a list of slot configuration for all encrypted devices in the system
+parseClevisConfig() {
+    luksDevices=$1
+    IFS="|"
+    for device in $luksDevices; do
+        logDebug "device=$device"
+        isReservedSlotPresent $device
+        isReserved=$?
+        pcrSlots=$(getPcrSlotsForDevice $device)
+        logDebug "pcrSlots=$pcrSlots"
+        parseClevisRegex "$pcrSlots" $isReserved "$device"
+    done
+}
+
+getPcrSlotsForDevice() {
+    device=$1
+    logDebug "getPcrSlotsForDevice, device=$device"
+    clevis luks list -d $device | grep -v "$RESERVED_SLOT:" | grep pcr_ids || true
+}
+
+parseClevisRegex() {
+    IFS=$'\n'
+    for line in $1; do
+        logDebug "line=$line"
+        echo "$line" | sed -E 's@([0-9]+)(:\s+.*+\s+'\'')(\{)(.*?"pcr_ids":")([^"]*)(".*)(.*)('\''.*)@'$2'|'$3'|\1|\5|\3\4\5\6\7@'
+    done
+}
+
+# executes a function pointer passed argument $1 for each slot configured with PCR and
+# for every device in the system
+processPCRentriesOnly() {
+    luksDevices=$(getLUKSDevices)
+    parsedClevis=$(parseClevisConfig $luksDevices)
+    if [ "$parsedClevis" == "" ]; then
+        logInfo "no pcr config detected, nothing to do for $1"
+        return
+    fi
+    logInfo "parsed clevis for all drives: $parsedClevis"
+    echo "$parsedClevis" | while IFS= read -r line; do
+        logDebug "$line"
+        IFS="|" read -ra values <<<"$line"
+        reservedSlotPresent=${values[0]}
+        device=${values[1]}
+        slotNumber=${values[2]}
+        pcrIDs=${values[3]}
+        clevisConfig=${values[4]}
+        logInfo "reservedSlot=$reservedSlotPresent device=$device slot=$slotNumber with PCR IDs=$pcrIDs and clevis config=$clevisConfig"
+        if [ -n "$pcrIDs" ]; then
+            logDebug "before applying command: $(/usr/bin/tpm2_pcrread sha256:$pcrIDs)"
+            $1 $reservedSlotPresent $device $slotNumber $pcrIDs $clevisConfig || true
```

Comment:
I'd rather you'll add that as a variable at the top of the function and then use the generic function name

Reply:
fixed